File: static/Scripts/faculty.js
let selected_subject;
document.addEventListener('DOMContentLoaded', function () {
    // Retrieve the user ID from the URL query parameter
    const urlParams = new URLSearchParams(window.location.search);
    const userID = urlParams.get('userID');

    // document.querySelector("wrapper").style.display="block";
    function fetchFacultyDetails() {
        fetch('/getFacultyDetails?userID=' + userID) // Replace with your actual endpoint to fetch faculty details
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch faculty details');
                }
            })
            .then(faculty => {
                // Populate the faculty details on the webpage
                document.getElementById('facultyName').innerText = faculty.facultyName;
                document.getElementById('facultyEmail').innerText = faculty.email;
                document.getElementById('facultyID').innerText = "ID: " + faculty.facultyId;
                // Add more fields as needed
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }

    function fetchSubjectButtons() {
        fetch('/getFacultySubjects?userID=' + userID) // Replace with your actual endpoint to fetch faculty subjects
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch faculty subjects');
                }
            })
            .then(subjects => {
                // Populate subject buttons on the webpage
                const nav = document.getElementById('subjectNav');
                subjects.forEach((subject, index) => {
                    const button = document.createElement('button');
                    button.id = `subject-${index}`;
                    button.textContent = subject.subject;
                    button.addEventListener('click', function () {
                        selected_subject = subject.subject;
                        let attendanceSection = document.querySelector(".attendance-section");
                        if (attendanceSection.style.display === "block") {
                            fetchStudentUserIDs(subject.subject);
                        }
                        let defaulter = document.querySelector(".defaulters");
                        if (defaulter.style.display === "block") {
                            fetchDefaulters(subject.subject);
                        }
                    });
                    nav.appendChild(button);
                });
                document.getElementById("subject-0").click();
            })
            .catch(error => {
                console.error('Error:', error);
                // Handle error, e.g., display error message to the user
            });
    }

    fetchFacultyDetails();
    setTimeout(fetchSubjectButtons, 1000);

    let userIDIndex = 0;
    let studentUserIDs = [];

    function fetchStudentUserIDs(subject) {
        fetch('/getStudentUserIDs?subject=' + subject)
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch student user IDs');
                }
            })
            .then(data => {
                studentUserIDs = data;
                updateStudentDetails();
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }

    function updateStudentDetails() {
        const userID = studentUserIDs[userIDIndex];
        fetch('/getStudentDetails?userID=' + userID)
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch student details');
                }
            })
            .then(student => {
                // Populate student details on the webpage
                document.getElementById('studentName').innerText = student.studentName;
                document.getElementById('rollNumber').innerText = student.studentId;
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }


    document.querySelector('.nextBtn').addEventListener('click', function () {
        userIDIndex = (userIDIndex + 1) % studentUserIDs.length;
        const previousMessageDiv = document.querySelector('.submit').nextElementSibling;
        if (previousMessageDiv) {
            previousMessageDiv.remove();
        }
        updateStudentDetails(); // Update student details
    });

    // Event listener for the "Back" button
    document.querySelector('.backBtn').addEventListener('click', function () {
        userIDIndex = (userIDIndex - 1 + studentUserIDs.length) % studentUserIDs.length; // Decrement the index
        const previousMessageDiv = document.querySelector('.submit').nextElementSibling;
        if (previousMessageDiv) {
            previousMessageDiv.remove();
        }
        updateStudentDetails(); // Update student details
    });
    document.querySelector('.submit').addEventListener('click', function () {
        const studentID = document.getElementById('rollNumber').innerText;

        //check if date is selected
        const date = document.getElementById('date').value;
        if (!date) {
            alert('Please select a date');
            return;
        }

        // Get the selected attendance status from the radio buttons
        let status;

        // Check which radio button is selected
        if (document.getElementById('present').checked) {
            status = 'Present';
        } else if (document.getElementById('absent').checked) {
            status = 'Absent';
        } else {
            // Handle the case where neither radio button is selected
            alert('Please select attendance status');
            return; // Stop execution if attendance status is not selected
        }

        const data = {
            'faculty_id': userID, 'student_id': studentID, 'subject': selected_subject, 'date': date, 'status': status
        };

        fetch('/markAttendance', {
            method: 'POST', headers: {
                'Content-Type': 'application/json'
            }, body: JSON.stringify(data)
        })
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to mark attendance');
                }
            })
            .then(data => {
                const previousMessageDiv = document.querySelector('.submit').nextElementSibling;
                if (previousMessageDiv) {
                    previousMessageDiv.remove();
                }
                const messageDiv = document.createElement('div');
                messageDiv.style.marginTop = '15px';
                messageDiv.style.color = 'green';
                messageDiv.textContent = data.message; // Assuming the response contains a 'message' field
                document.querySelector('.submit').insertAdjacentElement('afterend', messageDiv);
            })
            .catch(error => {
                console.error('Error:', error);
            });
    });

    function markAttendance() {

        document.querySelector(".mark").classList.add("active");
        document.querySelector(".leave").classList.remove("active");
        document.querySelector(".defaulter_button").classList.remove("active")
        document.querySelector(".logout").classList.remove("active");
        let leaveSection = document.querySelector(".leave-section");
        if (leaveSection.style.display === "block") {
            leaveSection.style.display = "none";
        }
        let attendanceSection = document.querySelector(".attendance-section");
        if (attendanceSection.style.display === "none") {
            document.querySelector(".nextBtn").style.display = "block";
            document.querySelector(".backBtn").style.display = "block";
            attendanceSection.style.display = "block";
        }
        let defaulter = document.querySelector(".defaulters");
        if (defaulter.style.display === "block") {
            defaulter.style.display = "none";
        }
        document.getElementById("subject-0").click();
    }

    function leaveApplication() {

        document.querySelector(".mark").classList.remove("active");
        document.querySelector(".leave").classList.add("active");
        document.querySelector(".defaulter_button").classList.remove("active")
        document.querySelector(".logout").classList.remove("active");
        let attendanceSection = document.querySelector(".attendance-section");
        if (attendanceSection.style.display === "block") {
            attendanceSection.style.display = "none";
            document.querySelector(".nextBtn").style.display = "none";
            document.querySelector(".backBtn").style.display = "none";
        }
        let leaveSection = document.querySelector(".leave-section");
        if (leaveSection.style.display === "none") {
            leaveSection.style.display = "block";
        }
        let defaulter = document.querySelector(".defaulters");
        if (defaulter.style.display === "block") {
            defaulter.style.display = "none";
        }
        fetchLeaveApplication();
    }

    function fetchLeaveApplication() {

        fetch('/getLeaveApplication?userID=' + userID)
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch leave application');
                }
            })
            .then(leaveApplication => {
                console.log(leaveApplication)
                const tableBody = document.querySelector('#leave_table tbody');
                if (tableBody) {
                    detachLeaveDetailsRow();
                    tableBody.innerHTML = '';
                } else {
                    console.error('Table body not found');
                }

                leaveApplication.forEach(application => {
                    console.log(application);
                    const tr = document.createElement('tr');
                    tr.innerHTML = `
                    <td>${application.applicationId}</td>
                    <td>${application.subject}</td>
                    <td>${application.studentId}</td>
                    <td>${application.studentName}</td>
                    <td>${application.status}</td>
                    `;
                    tr.addEventListener('click', function () {
                        reattachLeaveDetailsRow();
                        showLeaveDetails(application);
                    });
                    if (tableBody.firstChild) {
                        tableBody.insertBefore(tr, tableBody.firstChild);
                    } else {
                        tableBody.appendChild(tr);
                    }
                });
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }


    const leaveDetailsRow = document.getElementById('some');

    function detachLeaveDetailsRow() {
        if (leaveDetailsRow.parentNode) {
            leaveDetailsRow.parentNode.removeChild(leaveDetailsRow);
        }
    }

    function reattachLeaveDetailsRow() {
        const tableBody = document.querySelector('#leave_table tbody');
        if (tableBody) {
            tableBody.appendChild(leaveDetailsRow);
        }
    }


    function showLeaveDetails(application) {
        document.getElementById('leaveAppID').textContent = application.applicationId;
        document.getElementById('leaveSubject').textContent = application.subject;
        document.getElementById('leaveStudentID').textContent = application.studentId;
        document.getElementById('leaveStudentName').textContent = application.studentName;
        document.getElementById('leaveSDate').textContent = application.start_date.substring(0, 16);
        document.getElementById('leaveEDate').textContent = application.end_date.substring(0, 16);
        document.getElementById('leaveReason').textContent = application.reason;
        document.getElementById('leaveStatus').textContent = application.status;
        //  document.getElementById('leaveDetails').style.display = 'block';
        //  document.getElementById('leaveBtns').style.display = 'block';
        document.getElementById('some').style.display = "";
    }

    document.querySelector('.approveBtn').addEventListener('click', function () {
        approveLeave();
        setTimeout(fetchLeaveApplication, 1000);
    });

    document.querySelector('.rejectBtn').addEventListener('click', function () {
        rejectLeave();
        setTimeout(fetchLeaveApplication, 1000);
    });

    // Close the dropdown if the user clicks outside of it
    function defaulters(subject) {

        document.querySelector(".mark").classList.remove("active");
        document.querySelector(".leave").classList.remove("active");
        document.querySelector(".defaulter_button").classList.add("active")
        document.querySelector(".logout").classList.remove("active");
        let attendanceSection = document.querySelector(".attendance-section");
        if (attendanceSection.style.display === "block") {
            attendanceSection.style.display = "none";
            document.querySelector(".nextBtn").style.display = "none";
            document.querySelector(".backBtn").style.display = "none";
        }
        let leaveSection = document.querySelector(".leave-section");
        if (leaveSection.style.display === "block") {
            leaveSection.style.display = "none";
        }
        let defaulter = document.querySelector(".defaulters");
        if (defaulter.style.display === "none") {
            defaulter.style.display = "block";
        }
        document.getElementById("subject-0").click();
    }

    function approveLeave() {
        const applicationId = document.getElementById('leaveAppID').textContent;
        // Call API to approve the leave application with the provided applicationId
        fetch('/approveLeave', {
            method: 'POST', headers: {
                'Content-Type': 'application/json'
            }, body: JSON.stringify({applicationId})
        })
            .then(response => {
                if (response.ok) {
                    console.log('Leave application approved successfully');
                    // Optionally, update UI or perform other actions
                } else {
                    throw new Error('Failed to approve leave application');
                }
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }

    function rejectLeave() {
        const applicationId = document.getElementById('leaveAppID').textContent;
        // Call API to reject the leave application with the provided applicationId
        fetch('/rejectLeave', {
            method: 'POST', headers: {
                'Content-Type': 'application/json'
            }, body: JSON.stringify({applicationId})
        })
            .then(response => {
                if (response.ok) {
                    console.log('Leave application rejected successfully');
                    // Optionally, update UI or perform other actions
                } else {
                    throw new Error('Failed to reject leave application');
                }
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }

    function fetchDefaulters(subject) {
        document.querySelector(".selectSub").innerText = "";
        fetch('/getDefaulters?subject=' + subject)
            .then(response => {
                if (response.ok) {
                    return response.json();
                } else {
                    throw new Error('Failed to fetch attendance data');
                }
            })
            .then(attendance => {
                console.log(attendance)
                if (attendance.length == 0) {
                    console.log('table is empty')
                    document.querySelector(".emptyTable").style.display = "block";
                    document.querySelector("#defaulters_table").style.display = "none"
                } else {
                    document.querySelector(".emptyTable").style.display = "none";
                    document.querySelector("#defaulters_table").style.display = "table"
                }
                const tableBody = document.querySelector('#defaulters_table tbody');
                tableBody.innerHTML = '';
                attendance.forEach(record => {
                    console.log(record);
                    const tr = document.createElement('tr');
                    tr.innerHTML = `
                           <!--remove the time part-->
                        <td>${record[0]}</td>  
                        <td>${record[1]}</td>
                        <td>${record[2]}</td>
                        
                    `;
                    tableBody.appendChild(tr);
                });
            })
            .catch(error => {
                console.error('Error:', error);
            });
    }


    document.querySelector('.mark').addEventListener('click', function () {
        document.getElementById('subjectNav').style.display = "";
        markAttendance();
    });
    document.querySelector('.leave').addEventListener('click', function (e) {
        document.getElementById('subjectNav').style.display = "none";
        leaveApplication();
    });
    document.querySelector('.defaulter_button').addEventListener('click', function () {
        document.getElementById('subjectNav').style.display = "";
        defaulters();
    });
});
/*
document.querySelector('.mark').addEventListener('click', function () {

    document.getElementsByClassName("attendance-section").style.display = "none";
    document.getElementsByClassName("nextBtn").style.display = "visible";
    document.getElementsByClassName("backBtn").style.display = "visible";


});*/

File: app/faculty.py
from flask import Blueprint, request, jsonify, render_template

from app.connection import *

faculty_bp = Blueprint('faculty', __name__)


# Routes related to faculty functionality
@faculty_bp.route('/faculty')
def faculty_page():
    userID = request.args.get('userID')  # Retrieve user ID from query parameter
    return render_template('faculty.html', userID=userID)


@faculty_bp.route('/getFacultyDetails', methods=['GET'])
def get_faculty_details():
    userID = request.args.get('userID')
    print("UID; ", userID)
    cursor.execute("SELECT * FROM faculty WHERE facultyId = %s", (userID,))
    faculty = cursor.fetchall()
    faculty = faculty[0]
    return jsonify(faculty)


@faculty_bp.route('/getFacultySubjects', methods=['GET'])
def get_faculty_subjects():
    userID = request.args.get('userID')
    print(userID)
    cursor.execute("SELECT subject FROM faculty WHERE facultyId = %s", (userID,))
    subjects = cursor.fetchall()
    return jsonify(subjects)


@faculty_bp.route('/getStudentDetails', methods=['GET'])
def get_student_details():
    userID = request.args.get('userID')
    cursor.execute("SELECT * FROM student WHERE studentId = %s", (userID,))
    student = cursor.fetchall()
    student = student[0]
    return jsonify(student)


@faculty_bp.route('/getStudentUserIDs', methods=['GET'])
def get_student_user_ids():
    subject = request.args.get('subject')
    cursor.execute("SELECT studentId FROM student WHERE subject = %s", (subject,))
    results = cursor.fetchall()
    user_ids = [result['studentId'] for result in results]
    return jsonify(user_ids)


@faculty_bp.route('/markAttendance', methods=['POST'])
def mark_attendance():
    data = request.json
    student_id = data['student_id']
    faculty_id = data['faculty_id']
    subject = data['subject']
    date = data['date']
    status = data['status']

    # Check if the attendance record already exists
    cursor.execute("SELECT * FROM attendance WHERE subject = %s AND studentId = %s AND date = %s",
                   (subject, student_id, date))
    existing_record = cursor.fetchone()

    if existing_record:
        # If the record exists, update the status
        cursor.execute("UPDATE attendance SET status = %s WHERE subject = %s AND studentId = %s AND date = %s",
                       (status, subject, student_id, date))
        db.commit()
        return jsonify({'message': 'Attendance updated successfully'}), 200
    else:
        # If the record does not exist, insert a new record
        cursor.execute(
            "INSERT INTO attendance (subject, studentId, facultyId, date, status) VALUES (%s, %s, %s, %s, %s)",
            (subject, student_id, faculty_id, date, status))
        db.commit()
        return jsonify({'message': 'Attendance marked successfully'}), 200


@faculty_bp.route('/getLeaveApplication', methods=['GET'])
def get_leave_applications():
    userID = request.args.get('userID')
    cursor.execute("select * from leave_application where facultyId = %s", (userID,))
    applications = cursor.fetchall()
    for i in applications:
        cursor.execute("SELECT studentName FROM student WHERE studentId = %s", (i['studentId'],))
        i['studentName'] = cursor.fetchall()[0]['studentName']
    return jsonify(applications)


@faculty_bp.route('/approveLeave', methods=['POST'])
def approve_leave():
    # Get application ID from request body
    application_id = request.json['applicationId']
    cursor.execute("UPDATE leave_application SET status = 'Approved' WHERE applicationId = %s", (application_id,))
    # update the attendance table data if marked as absent
    cursor.execute("SELECT * FROM leave_application WHERE applicationId = %s", (application_id,))
    application = cursor.fetchall()[0]
    print(application)
    cursor.execute("UPDATE attendance SET status = 'Present' WHERE studentId = %s and status = 'Absent' "
                   "and date between %s and %s and subject = %s", (application['studentId'], application['start_date'],
                                                                   application['end_date'], application['subject']))

    db.commit()
    return jsonify({'message': 'Leave application approved successfully'}), 200


# Endpoint to reject leave application
@faculty_bp.route('/rejectLeave', methods=['POST'])
def reject_leave():
    # Get application ID from request body
    application_id = request.json['applicationId']
    cursor.execute("UPDATE leave_application SET status = 'Rejected' WHERE applicationId = %s", (application_id,))
    db.commit()
    return jsonify({'message': 'Leave application rejected successfully'}), 200


@faculty_bp.route('/getDefaulters', methods=['GET'])
def get_defaulters():
    subject = request.args.get('subject')
    attendance = []
    cursor.execute("SELECT studentId FROM student WHERE subject = %s", (subject,))
    user_ids = [result['studentId'] for result in cursor.fetchall()]
    cursor.execute("select attendencePercentageCriteria from faculty where subject = %s", (subject,))
    criterion = cursor.fetchall()[0]['attendencePercentageCriteria']
    for id in user_ids:
        cursor.execute(
            "SELECT COUNT(*) AS present_count FROM attendance WHERE studentId = %s AND subject = %s AND status = 'Present'",
            (id, subject))
        present_count = cursor.fetchall()[0]['present_count']
        cursor.execute("SELECT COUNT(*) AS total_count FROM attendance WHERE studentId = %s AND subject = %s",
                       (id, subject))
        total_count = cursor.fetchall()[0]['total_count']
        if total_count != 0:
            average_attendance = (present_count / total_count) * 100
        else:
            average_attendance = 0
        if average_attendance < criterion:
            temp = [id]
            cursor.execute("SELECT studentName FROM student WHERE studentId = %s", (id,))
            temp.append(cursor.fetchall()[0]['studentName'])
            average_attendance = round(average_attendance, 2)
            temp.append(average_attendance)
            attendance.append(temp)

    return jsonify(attendance)
